### PR TITLE
Pause OOP server when 'global operations' (i.e. build) are executing.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @heejaechang

**Comment by CyrusNajmabadi** [comment]
retest windows_debug_unit64_prtest  please

File: src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs [hunk 1]
```diff
@@ -31,6 +39,10 @@ internal sealed partial class ServiceHubRemoteHostClient : RemoteHostClient
         private readonly JsonRpc _rpc;
         private readonly ReferenceCountedDisposable<RemotableDataJsonRpc> _remotableDataRpc;
 
+        private readonly object _globalNotificationsGate = new object();
+        private Task<GlobalNotificationState> _globalNotificationsTask = Task.FromResult(GlobalNotificationState.NotStarted);
+        private KeepAliveSession _keepAliveSession;
```
Comment (heejaechang) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs]: @CyrusNajmabadi sorry about confusion. I think you don't actually need KeepAliveSession. Client already has _rpc that is connected to RemoteHostService. you can just use that one to send it over.

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

File: src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs]: Oh great!  That makes things much easier :)

**Comment by heejaechang** [comment]
looks good. I can think of a reason why it shouldn't be tied to client (sorry I just realized) but service since client can be restarted. but right now, that is experimental thing that is not exposed to outside, so seems fine.

File: src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs [hunk 2]
```diff
@@ -31,6 +39,10 @@ internal sealed partial class ServiceHubRemoteHostClient : RemoteHostClient
         private readonly JsonRpc _rpc;
         private readonly ReferenceCountedDisposable<RemotableDataJsonRpc> _remotableDataRpc;
 
+        private readonly object _globalNotificationsGate = new object();
+        private Task<GlobalNotificationState> _globalNotificationsTask = Task.FromResult(GlobalNotificationState.NotStarted);
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs]: Might be good to add a comment explaining what this serialization is attempting to address.

File: src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs [hunk 3]
```diff
@@ -31,6 +39,10 @@ internal sealed partial class ServiceHubRemoteHostClient : RemoteHostClient
         private readonly JsonRpc _rpc;
         private readonly ReferenceCountedDisposable<RemotableDataJsonRpc> _remotableDataRpc;
 
+        private readonly object _globalNotificationsGate = new object();
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs]: Comment which fields this is guarding.

File: src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs]: Done.

File: src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs]: Done

**Comment by CyrusNajmabadi** [comment]
@heejaechang Can you look again?

**Comment by heejaechang** [comment]
@CyrusNajmabadi @Pilchie once you check this in, I am thinking making Remote host not to be normal priority. I made it low priority to get similar effect as IGlobalOperationService does. now it has the service, it doesn't need to be lower priority.

other reason is that, I see a case where VS fail at start up due to it can't connect to OOP server since it is lower priority and machine is very busy (3 VS, 100% CPU being used), so connection time out and VS crash.

**Comment by CyrusNajmabadi** [comment]
retest windows_release_unit64_prtest please

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi @Pilchie once you check this in, I am thinking making Remote host not to be normal priority. I made it low priority to get similar effect as IGlobalOperationService does. now it has the service, it doesn't need to be lower priority.

Interesting.  My gut tells me to leave it at normal priority.  Just so that we're really not interfering with things like typing.  It's nice to have the OOP server causing as little contention as possible.

> other reason is that, I see a case where VS fail at start up due to it can't connect to OOP server since it is lower priority and machine is very busy (3 VS, 100% CPU being used), so connection time out and VS crash.

Feels like we might want to special case this.  i.e. start the server at normal priority so htat we can connect.  But then lower the priority so that it backs off.  Or maybe boost the priority every few seconds so it doesn't starve, but then lower for the remainder so it's non intrusive.

**Comment by heejaechang** [comment]
Sure, I can do that way then.

**Pull request merged**